openwisp_firmware_upgrader/tests/test_admin.py

```diff
@@ -319,6 +320,23 @@ def test_save_device_with_deleted_devicefirmware(self):
         )
         self.assertEqual(response.status_code, 200)
 
+    @capture_stderr()
+    def test_device_firmware_upgrade_without_device_connection(
+        self, captured_stderr, *args
+    ):
+        self._login()
+        device_fw = self._create_device_firmware()
+        device = device_fw.device
+        device.deviceconnection_set.all().delete()
+        response = self.client.get(
```

Comment:
Is the get request meant to trigger the error? I can't trigger the error by doing this.

Reply:
The device should not have any DeviceConnection object. This will not create a Server Error, but an error would be logged with logger.exception.